### ISSUE TITLE
Suggesting a Documentation Update: Basics and best practices for include statements, and the basics of how to split code across files

### ISSUE THREAD

**Issue by JamesKCS** [JuliaLang/julia#36799]
For context, see this reddit thread: https://www.reddit.com/r/Julia/comments/hwxpgm/beginner_question_files_modules_and_include_vs/  Everything below assumes that user a5sk6n's answer is the correct "best practice" answer.  If this is not the case, I think that this issue should still be better addressed in the docs, but, in that case, a lot of what I write below below will be wrong.

I (am some commenters on that thread) believe that the docs may not adequately address the basics of how to use include statements and how to split code across files.  This is something that (I think) should be addressed before before getting to the related but slightly more advanced topics of module and package creation.

Consider the following situation:

A beginner has main.jl, A.jl, B.jl, and C.jl. Assume:

-main.jl needs A.jl and B.jl and C.jl.

-A.jl needs B.jl and C.jl.

-B.jl needs C.jl.

The correct thing to do is only have include statements in main.jl, and nowhere else (assuming user a5sk6n's answer is the correct best practice).  In retrospect, this is obvious given the information in the docs.  However, many beginners tend to do what they would do in other languages: put include statements where they would be needed in a language like C++ or Python.  The result is subtle problems that come from running certain files (in this example, B.jl and C.jl) many times.  For me and at least several others, the blindly obvious correct approach above is only obvious after being told.

Frustratingly, not only do the docs not cover how to do this very basic thing correctly correctly, Internet searches for solutions tend to be quite unhelpful.  For example, threads and StackOverflow posts like this seem to be quite common: https://github.com/JuliaLang/julia/issues/30597.  I think what the OP in that thread needed was some basic guidance on include statements like what's above.  However, the answers immediately delved into module and package creation (not their fault, the OP was asking about that, but I suspect that those answers were not what he or she really needed).  Perhaps for the OP, and certainly for many others coming across that thread, what they really need is an answer like what's above, _before any more advanced stuff above package and module creation._

My suggestion is:
1) Early in the docs, before modules, have an section called something like "Organizing files and use of include statements: best practices".  This section would then has something like user a5sk6n's answer.  I've adapted this answer into a self-contained answer here: https://github.com/JuliaLang/julia/issues/30597#issuecomment-663812288 so copying my post there could be a good starting point.
2) In the “Modules and files” section of the docs, insert a reference and link referring back to the new section above.  Something along the lines of “for the basics of files and include statement best practices, see ... [reference back to the new section]”.

I think that many beginners coming from other languages share my confusion, and that these improvements to the docs would help a lot of people.  Improvements like this may also increase the popularity of Julia: when beginners get stuck on things like this, and a few searches don't help, I think a lot of people will give up on a language.

**Comment by Cvikli** [comment]
Me and 2 of my friend struggled with the same problem, also found the same situation, read every issue & comment on these threads but the include system best solution shows that you can have only 1 entry point of your program which is main.jl. Also it would be nice to have this advice("Organizing files and use of include statements: best practices") to have at the beginning when you read the docs. 
 
Sorry to ask on this thread, but what if some expert would create the ["include_once"](https://github.com/JuliaLang/julia/issues/30597#issue-396109817) macro. Then everyone who would have C++ or python background would be 100% satisfied with the include system of Julia? Also I am pretty sure if this include_once would use a "global include scope check", then it would solve the single entry point problem and [diamond problem at include](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) at once.

**Comment by jebej** [comment]
You just have to remember that ‘include’ is roughly the same as evaluating the code by hand. Everything is evaluated in main, unless you create modules. Order and dependency matters only for types, so a good way to work is to put your basic types in a file, and have other files grow on it.

Note that in the example above you could have: B includes C, A include B, and main include A.

**Comment by JamesKCS** [comment]
> You just have to remember that ‘include’ is roughly the same as evaluating the code by hand. Everything is evaluated in main, unless if you create modules.

Exactly! I knew this from reading the docs, and the solution to this behavior is blindingly obvious in retrospect.  However, before I knew the solution, the correct way of handling this was giving me problems: I was trying to do things the Python/C++ way, and that causes subtle problems.  Web search results (and that Reddit thread) reveal that I'm not the only one who struggled with this.

For this reason, I think it's worth explicitly spelling out not only the behavior of 'include', but a simple example for beginners showing the standard way of using these statements.

**Comment by jonas-schulze** [comment]
> I was trying to do things the Python/C++ way, and that causes subtle problems.

Maybe it’s helpful to think of the Julia `include(“Foo.jl”)` as the C `#include “foo.c”`, which will cause problems as well if you include the definitions twice. The Julia `using Foo` or `import Foo` correspond to the C `#include “foo.h”`, which is mostly fine to do twice, as long as you use include guards.

**Comment by aleferna12** [comment]
The problem that I see with this solution is that you will have trouble finding definitions in the files...

If B needs C, we can assume its structure is something like:

```
#B.jl

c = foo()
```

```
#C.jl

foo() = 1 + 1
```

But if I am reading file B.jl, how am I supposed to know where to find foo() if I don't explicitly include C in B?

This can get even more complicated if I want to find the definition of a function used in A, which needs both B and C. How do I know where to look for things in this system?
